avocado/core/nrunner.py

```diff
@@ -0,0 +1,232 @@
+import argparse
+import io
+import json
+import multiprocessing
+import subprocess
+import time
+import unittest
+
+#: The amount of time (in seconds) between each internal status check
+RUNNER_RUN_CHECK_INTERVAL = 0.01
+
+#: The amount of time (in seconds) between a status report from a
+#: runner that performs its work asynchronously
+RUNNER_RUN_STATUS_INTERVAL = 0.5
+
+
+class Runnable:
+    """
+    Describes an entity that be executed in the context of a task
+
+    A instance of :class:`BaseRunner` is the entity that will actually
+    execute a runnable.
+    """
+    def __init__(self, kind, uri=None, *args, **kwargs):
+        self.kind = kind
+        self.uri = uri
+        self.args = args
+        self.kwargs = kwargs
+
+
+def runnable_from_recipe(recipe_path):
+    """
+    Returns a runnable from a runnable recipe file
+    """
+    with open(recipe_path) as recipe_file:
+        recipe = json.load(recipe_file)
+    return Runnable(recipe.get('kind'),
+                    recipe.get('uri'),
+                    *recipe.get('args', ()))
+
+
+class BaseRunner:
+    """
+    Base interface for a Runner
+    """
+
+    def __init__(self, runnable):
+        self.runnable = runnable
+
+    def run(self):
+        yield {}
+
+
+class NoOpRunner(BaseRunner):
+    """
+    Sample runner that performs no action before reporting FINISHED status
+    """
+    def run(self):
+        yield {'status': 'finished'}
+
+
+class ExecRunner(BaseRunner):
+    """
+    Runner for standalone executables with or without arguments
+    """
+    def run(self):
+        process = subprocess.Popen(
+            [self.runnable.uri] + list(self.runnable.args),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE)
+
+        last_status = None
+        while process.poll() is None:
+            time.sleep(RUNNER_RUN_CHECK_INTERVAL)
+            now = time.time()
+            if last_status is None or now > last_status + RUNNER_RUN_STATUS_INTERVAL:
```

Comment:
I'm obsessed to optimize this conditional. :)

`last_status is None` will be evaluated `True` once only. If you set `last_status = - RUNNER_RUN_STATUS_INTERVAL` then in first iteration `now > 0` (i.e. last_status + RUNNER_RUN_STATUS_INTERVAL == - RUNNER_RUN_STATUS_INTERVAL + RUNNER_RUN_STATUS_INTERVAL== 0). Thus, this conditional becomes only `if now > last_status + RUNNER_RUN_STATUS_INTERVAL:`

Or am I missing something?